### PR TITLE
Change MiddlewareFactory to type activate IMiddleware as a fallback

### PR DESCRIPTION
**Pull request by davidfowl**
- This eases the transition and learning when you discover there's
an interface for middleware.
- Inspired by https://stackoverflow.com/questions/48394756/no-service-for-type-has-been-registered/
/cc @guardrex

### CODE REVIEW

**Comment by khellang** [comment]
This means you're able to activate `IMiddleware` instances without them being registered in the container first, i.e. by just doing `Use<TMiddleware>` where `TMiddlesware` implements `IMiddleware`?

**Comment by davidfowl** [comment]
@khellang yap

**Review by Tratcher — APPROVED** [review]

**Review by khellang — APPROVED** [review]

File: test/Microsoft.AspNetCore.Http.Abstractions.Tests/UseMiddlewareTest.cs [hunk 1]
```diff
@@ -232,6 +232,23 @@ public void UseMiddlewareWithIMiddlewareThrowsIfParametersSpecified()
             Assert.Same(middlewareFactory.Created, middlewareFactory.Released);
         }
 
+        [Fact]
+        public async Task UseMiddlewareWithIMiddlewareAndMiddlewareFactoryTypeActivates()
```
Comment (JunTaoLuo) [inline test/Microsoft.AspNetCore.Http.Abstractions.Tests/UseMiddlewareTest.cs]: How is this test different from the one above?

File: test/Microsoft.AspNetCore.Http.Abstractions.Tests/UseMiddlewareTest.cs [hunk 2]
```diff
@@ -232,6 +232,23 @@ public void UseMiddlewareWithIMiddlewareThrowsIfParametersSpecified()
...
+        {
+            var mockServiceProvider = new DummyServiceProvider();
+            var builder = new ApplicationBuilder(mockServiceProvider);
+            builder.UseMiddleware(typeof(Middleware));
+            var app = builder.Build();
+            var context = new DefaultHttpContext();
+            var sp = new DummyServiceProvider();
+            var middlewareFactory = new MiddlewareFactory(sp);
```
Comment (JunTaoLuo) [inline test/Microsoft.AspNetCore.Http.Abstractions.Tests/UseMiddlewareTest.cs]: Nevermind, it uses MiddlewareFactory instead of BasicMiddlewareFactory

**Review by JunTaoLuo — APPROVED** [review]

**Pull request merged**